### PR TITLE
Update Mastodon to Rails 6.1

### PR DESCRIPTION
**Pull request by ClearlyClaire**
PRs to merge first:
- [x] Prepare Mastodon for Rails 6 (#15911)
- [x] Improve account counters handling (#15913)
- [x] Further preparation for Rails 6 (#15916)
- [x] Prepare Mastodon for zeitwerk autoloader (#15917)

PRs to consider first:
- [x] Use more robust hook for loading timestamp_id function into database (#15919)

This is work in progress, some tests are still failing:
- [x] the code in `increment_count!` to handle race conditions isn't working correctly. Addressed by #15913
- [ ] the tests about the `force_ssl` config do not work anymore, i am not sure how to test those
- [x] migrations still fail (partly because of https://github.com/rails/rails/issues/41693, worked around)

Other broken things to look fix:
- [x] the hooks surrounding database setup seem to have changed somehow. Addressed by #15919

Nice to have:
- [x] enable zeitwerk (new autoloading code, more deterministic and supposedly faster): see #15917

This PR also does not switch to the new framework defaults yet, this means, among other things, that zeitwerk isn't enabled yet, and will require more work.

### CODE REVIEW

**Comment by ClearlyClaire** [comment]
To elaborate on the “hooks surrounding database setup seem to have changed somehow”, when fiddling with my test setup, I often get the following:

> PG::UndefinedFunction: ERROR:  function timestamp_id(text) does not exist
LINE 1: ...E TABLE "encrypted_messages" ("id" bigint DEFAULT timestamp_...

This doesn't seem to occur in the CI, but still, something is off and I need to find what before this PR is ready.

**Comment by ClearlyClaire** [comment]
Marked as ready for review, as I'm fairly happy with how it looks now, with the following caveats:
- it hasn't been tested in a production environment yet
- I'm not super happy with having to switch with unreleased forks for a few gems, but they are currently unmaintained

**Comment by ClearlyClaire** [comment]
I have found no regression so far, but clearing the Rails cache is required as some of the stored objects are Rails models that don't load quite the same on Rails 6.1.

File: config/environments/production.rb [hunk 1]
```diff
@@ -44,6 +44,13 @@
   # Allow to specify public IP of reverse proxy if it's needed
   config.action_dispatch.trusted_proxies = ENV['TRUSTED_PROXY_IP'].split.map { |item| IPAddr.new(item) } if ENV['TRUSTED_PROXY_IP'].present?
 
+  config.force_ssl = true
```
Comment (Gargron) [inline config/environments/production.rb:47]: I remember there were some recent Tor-related changes that mentioned force_ssl, does this conflict with that?

**Review by Gargron — APPROVED** [review]

**Comment by zunda** [comment]
As [ClearlyClaire wrote](https://github.com/tootsuite/mastodon/pull/15910#issuecomment-802923945), this change seems to make the data in Rails' cache inconsistent with the code resulting with errors like below. Clearing the cache with

```
tootctl cache clear
```

seems to have resolved the inconsistency.

```
method=GET path=/users/zundan/statuses/105946269594904129 format=json controller=StatusesController action=show status=500 error='ActionController::UrlGenerationError: No route matches {:action=>"show", :controller=>"emojis", :id=>#<CustomEmoji id: 183, shortcode: "saba", domain: nil, image_file_name: "food_kandume_saba.png", image_content_type: "image/png", image_file_size: 49717, image_updated_at: "2017-10-07 00:27:30.774446000 +0000", created_at: "2017-09-29 01:28:38.632698000 +0000", updated_at: "2017-10-07 00:27:31.124663000 +0000", disabled: false, uri: nil, image_remote_url: nil, visible_in_picker: true, category_id: nil, image_storage_schema_version: nil>}, possible unmatched constraints: [:id]
Did you mean?  emoji_path
new_admin_custom_emoji_url
new_admin_custom_emoji_path
batch_admin_custom_emojis_url' duration=519.68 view=0.00 db=224.12 key=acct:ClworldF@md.ggtea.org

ActionController::UrlGenerationError (No route matches {:action=>"show", :controller=>"emojis", :id=>#<CustomEmoji id: 183, shortcode: "saba", domain: nil, image_file_name: "food_kandume_saba.png", image_content_type: "image/png", image_file_size: 49717, image_updated_at: "2017-10-07 00:27:30.774446000 +0000", created_at: "2017-09-29 01:28:38.632698000 +0000", updated_at: "2017-10-07 00:27:31.124663000 +0000", disabled: false, uri: nil, image_remote_url: nil, visible_in_picker: true, category_id: nil, image_storage_schema_version: nil>}, possible unmatched constraints: [:id]
Did you mean?  emoji_path
new_admin_custom_emoji_url
new_admin_custom_emoji_path
batch_admin_custom_emojis_url):

app/lib/activitypub/tag_manager.rb:43:in `uri_for'
app/serializers/activitypub/emoji_serializer.rb:13:in `id'
app/lib/activitypub/serializer.rb:36:in `serializable_hash'
app/lib/activitypub/serializer.rb:36:in `serializable_hash'
app/lib/activitypub/adapter.rb:42:in `serializable_hash'
app/controllers/concerns/cache_concern.rb:22:in `render_with_cache'
app/controllers/statuses_controller.rb:38:in `block (2 levels) in show'
app/controllers/statuses_controller.rb:29:in `show'
```
```

**Pull request merged**